src/platform/i386/chal/chal_proto.h

```diff
@@ -41,21 +45,43 @@ struct cap_pgtbl {
 	struct cap_header h;
 	u32_t             refcnt_flags; /* includes refcnt and flags */
 	pgtbl_t           pgtbl;
-	u32_t             lvl;       /* what level are the pgtbl nodes at? */
-	struct cap_pgtbl *parent;    /* if !null, points to parent cap */
-	u64_t             frozen_ts; /* timestamp when frozen is set. */
+	u32_t             lvl;          /* what level are the pgtbl nodes at? */
+	asid_t            asid;         /* hardware context identifier */
+	struct cap_pgtbl *parent;       /* if !null, points to parent cap */
+	u64_t             frozen_ts;    /* timestamp when frozen is set. */
 } __attribute__((packed));
 
+extern pgtbl_t tlb_asid_active[NUM_CPU][NUM_ASID_MAX];
+
 /* Update the page table */
 static inline void
 chal_pgtbl_update(struct pgtbl_info *pt)
-{
-	asm volatile("mov %0, %%cr3" : : "r"(pt->pgtbl));
+{	
+	pgtbl_t *curr_cached = &tlb_asid_active[get_cpuid()][pt->asid];
+
+	/* currently cached pgtbl is this pgtbl */
+	if (*curr_cached == pt->pgtbl) {
+		__writecr3((unsigned long)pt->pgtbl | pt->asid | CR3_NO_FLUSH);
+		return;
+	}
+
+	/* no pgtbl cached for this asid */
+	if (*curr_cached == 0) {
```

Comment:
Might not be worth having this case, TBH. Without this case, we'll have an additional TLB flush only for the first time we switch to a new asid. That will happen max 4096 (or whatever the # of PCIDs is) times in the system. Likely not a worthwhile optimization.